### PR TITLE
Plane: add climb-before-turn on RTL

### PR DESCRIPTION
**Pull request by skorokithakis**
This PR adds a mode to ArduPilot that climbing *before* turning on RTL. This is meant to avoid accidents, as the least likely place to have an obstacle when flying is in front of the craft (and up). Thus, this climbs to the safe altitude (`ALT_HOLD_RTL`) before initiating a turn.

I'm very new to the codebase, so any tips/feedback would be appreciated. Specifically, I'm not sure how to properly add the parameter.

### CODE REVIEW

**Comment by IamPete1** [comment]
Thanks for the PR!

I'm not sure how this is different than the existing RTL_CLIMB_MIN with LEVEL_ROLL_LIMIT = 0? Or do you need LEVEL_ROLL_LIMIT at some none 0 value for take off and landing?

**Comment by skorokithakis** [comment]
Just today I had an issue with `RTL_CLIMB_MIN`: I was launching from a cliff down to a lake below, and when hitting RTL, the craft would climb by 30 meters (the maximum with `RTL_CLIMB_MIN`) and turn around and attempt to smash into the side of the cliff.

Since it started at -60 meters, the safe thing to do would be to climb to 100 meters (my RTL altitude) first (ie over the lake) and then turn around and come back to me. This PR ensures that the craft will clear any obstacle first, by reaching a safe altitude, before initiating a turn.

**Comment by skorokithakis** [comment]
@IamPete1 if you don't want to merge the PR, could you at least let me know if `plane.current_loc.alt > plane.next_WP_loc.alt` is the right thing to check in this instance, so I can run the code on my planes? Messing with RTL is always risky, especially when you don't know the codebase.

**Comment by skorokithakis** [comment]
As for the banking change, I don't understand why it was set to `LEVEL_ROLL_LIMIT` rather than going straight ahead, so I set it to 1 for this PR. I didn't know whether I could safely set it to 0, and 1 worked in the SITL so that's what I went with.

**Comment by IamPete1** [comment]
Ah, I see, its a above home vs above current location thing. I think it would probably be fair just to change that globally. So https://github.com/ArduPilot/ardupilot/blob/bdeadce95bbec83a9367c8a83c43fe72dc145b52/ArduPlane/mode_rtl.cpp#L27 

Would become

`        if (!plane.rtl.done_climb && (plane.current_loc.alt - MAX(plane.next_WP_loc.alt,plane.prev_WP_loc.alt))*0.01 > plane.g2.rtl_climb_min) {
`

**Comment by skorokithakis** [comment]
Does that mean "If the current location is higher than rtl_climb_min plus the max of the starting altitude and the ending altitude, stop the climb"?

I'm not sure I understand the purpose, wouldn't that always climb to `ALT_HOLD_RTL + RTL_CLIMB_MIN`?

**Comment by skorokithakis** [comment]
In general, it seems to me that `RTL_CLIMB_MIN` was intended to do what "climb before turn" does, but I don't see any occasion in which `RTL_CLIMB_MIN` is better than "climb before turn". If you're much lower than `ALT_HOLD_RTL`, it won't climb enough, and if you're higher, it wastes energy climbing even though you're already at a safe altitude, so "climb before turn" seems to me like it would be superior in all cases.

I'm probably not parsing your proposal correctly, though, so I can't tell whether that's better.

**Comment by skorokithakis** [comment]
Oh, I think I see what you propose, so we climb `RTL_CLIMB_MIN` meters above either our current altitude, or `ALT_HOLD_RTL`, whichever is highest? So if the craft is already over the RTL alt, it will only climb for a bit and then return?

That's a good solution, I wouldn't want to be climbing much if I'm already over `ALT_HOLD_RTL` but I can just set `RTL_CLIMB_MIN` to 1 and get basically the same behaviour, no?

**Comment by skorokithakis** [comment]
I have implemented your suggestion, as it seems easier. I have set the roll limit to `-1`/`1` because I'm not sure why doing a very wide circle would be preferable when climbing, but I can revert it if there's a reason.

File: ArduPlane/mode_rtl.cpp [hunk 1]
```diff
@@ -21,17 +21,17 @@ void ModeRTL::update()
...
     if (plane.g2.rtl_climb_min > 0) {
         /*
-          when RTL first starts limit bank angle to LEVEL_ROLL_LIMIT
-          until we have climbed by RTL_CLIMB_MIN meters
+          when RTL first starts limit bank angle until we have climbed
+          by RTL_CLIMB_MIN meters above RTL start or ALT_HOLD_RTL,
+          whichever is highest.
          */
-        if (!plane.rtl.done_climb && (plane.current_loc.alt - plane.prev_WP_loc.alt)*0.01 > plane.g2.rtl_climb_min) {
+        if (!plane.rtl.done_climb && (plane.current_loc.alt > MAX(plane.next_WP_loc.alt,plane.prev_WP_loc.alt))*0.01 + plane.g2.rtl_climb_min) {
```
Comment (IamPete1) [inline ArduPlane/mode_rtl.cpp:28]: In moving the calculation you have lost the conversion of rtl_climb_min from M to cm
```suggestion
if (!plane.rtl.done_climb && (plane.current_loc.alt - MAX(plane.next_WP_loc.alt,plane.prev_WP_loc.alt))*0.01 > plane.g2.rtl_climb_min) {
```

File: ArduPlane/mode_rtl.cpp [hunk 2]
```diff
@@ -45,7 +45,7 @@ void ModeRTL::navigate()
 
     if (plane.g.rtl_autoland == 1 &&
         !plane.auto_state.checked_for_autoland &&
-        plane.reached_loiter_target() && 
+        plane.reached_loiter_target() &&
```
Comment (IamPete1) [inline ArduPlane/mode_rtl.cpp:48]: this is unrelated white space

File: ArduPlane/mode_rtl.cpp [hunk 3]
```diff
@@ -21,17 +21,17 @@ void ModeRTL::update()
...
-        if (!plane.rtl.done_climb && (plane.current_loc.alt - plane.prev_WP_loc.alt)*0.01 > plane.g2.rtl_climb_min) {
+        if (!plane.rtl.done_climb && (plane.current_loc.alt > MAX(plane.next_WP_loc.alt,plane.prev_WP_loc.alt))*0.01 + plane.g2.rtl_climb_min) {
             plane.prev_WP_loc = plane.current_loc;
             plane.setup_glide_slope();
             plane.rtl.done_climb = true;
         }
         if (!plane.rtl.done_climb) {
-            plane.roll_limit_cd = MIN(plane.roll_limit_cd, plane.g.level_roll_limit*100);
-            plane.nav_roll_cd = constrain_int32(plane.nav_roll_cd, -plane.roll_limit_cd, plane.roll_limit_cd);
+            plane.nav_roll_cd = constrain_int32(plane.nav_roll_cd, -1, 1);
```
Comment (IamPete1) [inline ArduPlane/mode_rtl.cpp:34]: I think its better to leave with the level roll limit, you can then just change the param to 0 if you want 0. 

This current change might as well be 
```suggestion
            plane.nav_roll_cd = 0.0f;
```

Comment (skorokithakis) [reply ArduPlane/mode_rtl.cpp:34]: This is true, what I'm unsure of is this part of the docs:

> Setting this to zero will completely disable heading hold on auto takeoff and final landing approach.

Given that, I wasn't sure what the impact of setting this to zero would be to RTL.

File: ArduPlane/mode_rtl.cpp (hunk 2 above)
Comment (skorokithakis) [reply ArduPlane/mode_rtl.cpp:48]: Will fix, my editor automatically trims trailing whitespace.

File: ArduPlane/mode_rtl.cpp (hunk 1 above)
Comment (skorokithakis) [reply ArduPlane/mode_rtl.cpp:28]: Oops, thanks.

**Comment by skorokithakis** [comment]
Feedback addressed, thanks! Though I'm still not sure of the purpose of doing a slow turn while climbing to avoid obstacles, it seems to me that you'd want to climb perfectly straight until you were at a safe altitude.

**Comment by peterbarker** [comment]
I've marked this as DevCall for discussion, but haven't looked at it myself.

I'd be happier if there was an autotest that failed before and passed afterwards, to highlight the change.

**Comment by skorokithakis** [comment]
Hmm, I'm entirely unsure on how to run (or even write) an autotest, unfortunately...

**Review by tridge — CHANGES_REQUESTED** [review]
we cannot make this change unconditional. It sounds great for the example of flying off a cliff, but for other examples this change will be awful and could lead to a flyaway.
Take the example of flying on perfectly flat ground, and RTL happens due to battery failsafe while 1km away and flying away from the pilot. Battery is low, so will have difficulty climbing. With this change it will try to climb to the RTL alt before turning. That will take it a lot further away from the pilot and could take a long time (may in fact never reach the desired alt, as battery is low).

**Comment by skorokithakis** [comment]
@tridge Would a "climb before turn" boolean param work?

Also, what is the rationale behind doing a large turn at `LEVEL_ROLL_LIMIT` rather than setting roll to 0 while climbing?

**Comment by tridge** [comment]
a FLIGHT_OPTIONS bit would be good

**Comment by skorokithakis** [comment]
@tridge requested changes implemented and tested in the SITL.

**Comment by skorokithakis** [comment]
What's the next step for this?

File: ArduPlane/mode_rtl.cpp [hunk 4]
```diff
@@ -19,7 +19,21 @@ void ModeRTL::update()
     plane.calc_nav_pitch();
     plane.calc_throttle();
 
-    if (plane.g2.rtl_climb_min > 0) {
+    if (plane.g2.flight_options & FlightOptions::CLIMB_BEFORE_TURN) {
```
Comment (magicrub) [inline ArduPlane/mode_rtl.cpp:22]: this is a whole lot of redundant logic which makes it harder to read and maintain. All that is changing is the altitude threshold. Perhaps this could get reduced to:
```
    bool alt_threshold_reached = false;
    if (plane.g2.flight_options & FlightOptions::CLIMB_BEFORE_TURN) {
        // Climb to ALT_HOLD_RTL before turning. This overrides RTL_CLIMB_MIN.
        alt_threshold_reached = plane.current_loc.alt > plane.next_WP_loc.alt
    } else if (plane.g2.rtl_climb_min > 0) {
        /*
          when RTL first starts limit bank angle to LEVEL_ROLL_LIMIT
          until we have climbed by RTL_CLIMB_MIN meters
         */
        alt_threshold_reached = (plane.current_loc.alt - plane.prev_WP_loc.alt)*0.01 > plane.g2.rtl_climb_min
    } else {
        return;
    }

    if (!plane.rtl.done_climb && alt_threshold_reached) {
            plane.prev_WP_loc = plane.current_loc;
            plane.setup_glide_slope();
            plane.rtl.done_climb = true;
        }
    if (!plane.rtl.done_climb) {
            // Constrain the roll limit as a failsafe, that way if something goes wrong the plane will
            // eventually turn back and go to RTL instead of going perfectly straight. This also leaves
            // some leeway for fighting wind.
            plane.roll_limit_cd = MIN(plane.roll_limit_cd, plane.g.level_roll_limit*100);
            plane.nav_roll_cd = constrain_int32(plane.nav_roll_cd, -plane.roll_limit_cd, plane.roll_limit_cd);
    }




```

Comment (skorokithakis) [reply ArduPlane/mode_rtl.cpp:22]: Ah, you are correct, of course. I will change it shortly.

**Comment by skorokithakis** [comment]
Changes implemented. @tridge does this look good to merge?

**Comment by skorokithakis** [comment]
I added an autotest too, just to prevent any regressions later.

**Comment by Hwurzburg** [comment]
Probably needs some flight testing....I can do it later this weekend....

**Comment by skorokithakis** [comment]
@Hwurzburg  I have already flight-tested it and it works as expected on a F405-Wing, but of course additional testing would be appreciated.

**Comment by skorokithakis** [comment]
@Hwurzburg testing this again today, it works fine, but the plane levels off just before reaching RTL alt and takes a while to reach the last meter. I've added a check to push the TECS to climb if CLIMB_BEFORE_TURN is set, like RTL_CLIMB_MIN does.

**Comment by skorokithakis** [comment]
Alright, just finished flying this again, it works *beautifully*. Climbed to RTL alt quickly and started to come back accurately every time. This is a fantastic feature, I feel much safer flying this way. I strongly, strongly recommend this makes the next release, if possible.

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Just got a failure on a PR against master:
    ClimbBeforeTurn (Test climb-before-turn) (Expected to be closer to  home (was 534.503104m, now 536.883549m).) (see /tmp/buildlogs/ArduPlane-ClimbBeforeTurn.txt)

Could we make sure this new test reliably passes, please?

**Comment by skorokithakis** [comment]
@peterbarker certainly. Any suggestions as to how?

**Comment by peterbarker** [comment]
On Tue, 16 Mar 2021, Stavros Korokithakis wrote:

> @peterbarker certainly. Any suggestions as to how?

Generally loosen the constraints until when you run it 10 times in a row 
it doesn't fail.

If the test is not useful after that then you have to get more clever 
about what you're testing - make the change in the thing you're looking at 
more dramatic, for example.

**Comment by skorokithakis** [comment]
@peterbarker thanks, should I push to this PR again or make a new one?

**Comment by skorokithakis** [comment]
I've addressed this in #16916.

**Comment by peterbarker** [comment]
On Tue, 16 Mar 2021, Stavros Korokithakis wrote:

> @peterbarker thanks, should I push to this PR again or make a new one?

New one, please.
